### DISCUSSION TITLE
NoSuchMethodError During PDF Export with PrimeFaces 14.0.3

### DISCUSSION THREAD

**Discussion by buddhika75** [primefaces/community#3047] [PrimeFaces] [answered]
I am encountering an issue while exporting a DataTable to PDF using PrimeFaces 14.0.3. The following error occurs:

```
java.lang.NoSuchMethodError: 'com.lowagie.text.pdf.PdfPCell com.lowagie.text.pdf.PdfPTable.addCell(com.lowagie.text.pdf.PdfPCell)'
```

**Environment Details:**

- PrimeFaces Version: 14.0.3 (also tried 14.0.6)
- OpenPDF Version: 1.4.1 (also tried 1.4.2)
- JSF Version: 2.3
- Java Version: 11
- Application Server: Payara 5

**Relevant `pom.xml` Dependencies:**

```xml
<dependency>
    <groupId>com.github.librepdf</groupId>
    <artifactId>openpdf</artifactId>
    <version>1.4.1</version>
</dependency>

<dependency>
    <groupId>org.primefaces</groupId>
    <artifactId>primefaces</artifactId>
    <version>14.0.3</version>
</dependency>
```

**Steps to Reproduce:**

1. Implement a DataTable in a JSF page.
2. Add a `p:commandButton` with a nested `p:dataExporter` for PDF export.
3. Deploy the application and attempt to export the DataTable as a PDF.

**Observed Behavior:**

Upon initiating the PDF export, the application throws the `NoSuchMethodError` mentioned above.

**Troubleshooting Steps Taken:**

- Verified that the OpenPDF and PrimeFaces versions are compatible as per the [PrimeFaces documentation](https://www.primefaces.org/docs/v14.0.3/).
- Cleaned and rebuilt the project to ensure all dependencies are correctly applied.
- Reviewed the `pom.xml` for potential conflicts with other dependencies.

**Request for Assistance:**

I am seeking guidance on resolving this `NoSuchMethodError`. Specifically:

- Is there a known compatibility issue between PrimeFaces 14.0.3 and OpenPDF 1.4.1 for PDF exports?
- Could other dependencies in my project be interfering with the PDF generation process?
- Are there additional configurations required in PrimeFaces or OpenPDF to handle PDF exports correctly?

Any insights or recommendations to address this issue would be greatly appreciated.

**Comment by melloware** [comment]
Weird the PF showcase uses 1.4.2 and works fine?

```xml
<dependency>
                <groupId>com.github.librepdf</groupId>
                <artifactId>openpdf</artifactId>
                <version>1.4.2</version>
            </dependency>
```

**Reply by melloware** [reply]
can you do `mvn depednency:tree` and look for iText.  

I wonder if you have a legacy iText JAR somehow in the way that is why it can't find  `java.lang.NoSuchMethodError: 'com.lowagie.text.pdf.PdfPCell com.lowagie.text.pdf.PdfPTable.addCell(com.lowagie.text.pdf.PdfPCell)'` ???

**Reply by buddhika75** [reply]
> can you do `mvn depednency:tree` and look for iText.
> 
> I wonder if you have a legacy iText JAR somehow in the way that is why it can't find `java.lang.NoSuchMethodError: 'com.lowagie.text.pdf.PdfPCell com.lowagie.text.pdf.PdfPTable.addCell(com.lowagie.text.pdf.PdfPCell)'` ???

 --- maven-dependency-plugin:3.1.2:tree (default-cli) @ rhdev ---
[INFO] com.divudi:rhdev:war:3.0.0
[INFO] +- ca.uhn.hapi.fhir:hapi-fhir-base:jar:6.1.3:compile
[INFO] |  +- com.fasterxml.jackson.datatype:jackson-datatype-jsr310:jar:2.13.2:compile
[INFO] |  +- org.apache.commons:commons-text:jar:1.9:compile
[INFO] |  +- commons-codec:commons-codec:jar:1.15:compile
[INFO] |  +- com.google.guava:guava:jar:31.0.1-jre:compile
[INFO] |  |  +- com.google.guava:failureaccess:jar:1.0.1:compile
[INFO] |  |  +- com.google.guava:listenablefuture:jar:9999.0-empty-to-avoid-conflict-with-guava:compile
[INFO] |  |  +- org.checkerframework:checker-qual:jar:3.12.0:compile
[INFO] |  |  +- com.google.errorprone:error_prone_annotations:jar:2.7.1:compile
[INFO] |  |  \- com.google.j2objc:j2objc-annotations:jar:1.3:compile
[INFO] |  +- org.slf4j:slf4j-api:jar:1.7.33:compile
[INFO] |  +- org.slf4j:jcl-over-slf4j:jar:1.7.33:compile
[INFO] |  \- com.google.code.findbugs:jsr305:jar:3.0.2:compile
[INFO] +- ca.uhn.hapi.fhir:hapi-fhir-structures-r5:jar:6.1.3:compile
[INFO] |  +- com.squareup.okhttp3:okhttp:jar:4.10.0:compile
[INFO] |  |  +- com.squareup.okio:okio-jvm:jar:3.0.0:compile
[INFO] |  |  |  +- org.jetbrains.kotlin:kotlin-stdlib-jdk8:jar:1.5.31:compile
[INFO] |  |  |  |  \- org.jetbrains.kotlin:kotlin-stdlib-jdk7:jar:1.5.31:compile
[INFO] |  |  |  \- org.jetbrains.kotlin:kotlin-stdlib-common:jar:1.5.31:compile
[INFO] |  |  \- org.jetbrains.kotlin:kotlin-stdlib:jar:1.6.20:compile
[INFO] |  |     \- org.jetbrains:annotations:jar:13.0:compile
[INFO] |  +- ca.uhn.hapi.fhir:org.hl7.fhir.utilities:jar:5.6.36:compile
[INFO] |  \- ca.uhn.hapi.fhir:org.hl7.fhir.r5:jar:5.6.36:compile
[INFO] +- ca.uhn.hapi:hapi-structures-v25:jar:2.3:compile
[INFO] +- ca.uhn.hapi:hapi-base:jar:2.3:compile
[INFO] +- ca.uhn.hapi:hapi-structures-v24:jar:2.3:compile
[INFO] +- org.codehaus.jackson:jackson-jaxrs:jar:1.9.13:compile
[INFO] |  +- org.codehaus.jackson:jackson-core-asl:jar:1.9.13:compile
[INFO] |  \- org.codehaus.jackson:jackson-mapper-asl:jar:1.9.13:compile
[INFO] +- com.fasterxml.jackson.core:jackson-databind:jar:2.14.1:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.14.1:compile
[INFO] |  \- com.fasterxml.jackson.core:jackson-core:jar:2.14.1:compile
[INFO] +- commons-beanutils:commons-beanutils:jar:1.9.4:compile
[INFO] |  \- commons-collections:commons-collections:jar:3.2.2:compile
[INFO] +- commons-io:commons-io:jar:2.15.0:compile
[INFO] +- org.apache.commons:commons-lang3:jar:3.14.0:compile
[INFO] +- commons-fileupload:commons-fileupload:jar:1.5:compile
[INFO] +- commons-logging:commons-logging:jar:1.1:compile
[INFO] |  +- log4j:log4j:jar:1.2.12:compile
[INFO] |  +- logkit:logkit:jar:1.0.1:compile
[INFO] |  +- avalon-framework:avalon-framework:jar:4.1.3:compile
[INFO] |  \- javax.servlet:servlet-api:jar:2.3:compile
[INFO] +- joda-time:joda-time:jar:2.12.4:compile
[INFO] +- org.glassfish:javax.el:jar:3.0.0:compile
[INFO] +- org.glassfish.web:el-impl:jar:2.2:compile
[INFO] |  \- javax.el:el-api:jar:2.2:compile
[INFO] +- com.google.zxing:core:jar:3.4.1:compile
[INFO] +- org.apache.poi:poi:jar:5.2.5:compile
[INFO] |  +- org.apache.commons:commons-collections4:jar:4.4:compile
[INFO] |  +- org.apache.commons:commons-math3:jar:3.6.1:compile
[INFO] |  +- com.zaxxer:SparseBitSet:jar:1.3:compile
[INFO] |  \- org.apache.logging.log4j:log4j-api:jar:2.21.1:compile
[INFO] +- org.apache.poi:poi-ooxml:jar:5.2.5:compile
[INFO] |  +- org.apache.poi:poi-ooxml-lite:jar:5.2.5:compile
[INFO] |  +- org.apache.xmlbeans:xmlbeans:jar:5.2.0:compile
[INFO] |  \- org.apache.commons:commons-compress:jar:1.25.0:compile
[INFO] +- fr.opensagres.xdocreport:fr.opensagres.poi.xwpf.converter.xhtml:jar:2.0.4:compile
[INFO] |  +- fr.opensagres.xdocreport:fr.opensagres.poi.xwpf.converter.core:jar:2.0.4:compile
[INFO] |  |  \- org.apache.poi:poi-ooxml-full:jar:5.2.0:compile
[INFO] |  \- fr.opensagres.xdocreport:fr.opensagres.xdocreport.core:jar:2.0.4:compile
[INFO] +- com.thoughtworks.xstream:xstream:jar:1.4.20:compile
[INFO] |  \- io.github.x-stream:mxparser:jar:1.2.2:compile
[INFO] |     \- xmlpull:xmlpull:jar:1.1.3.1:compile
[INFO] +- net.sourceforge.barbecue:barbecue:jar:1.5-beta1:compile
[INFO] +- mysql:mysql-connector-java:jar:8.0.30:compile
[INFO] |  \- com.google.protobuf:protobuf-java:jar:3.19.4:compile
[INFO] +- com.github.librepdf:openpdf:jar:1.3.24:compile
[INFO] +- org.apache.logging.log4j:log4j-core:jar:3.0.0-alpha1:compile
[INFO] |  +- org.apache.logging.log4j:log4j-plugins:jar:3.0.0-alpha1:compile
[INFO] |  +- javax.activation:javax.activation-api:jar:1.2.0:runtime
[INFO] |  +- javax.xml.bind:jaxb-api:jar:2.3.1:runtime
[INFO] |  \- com.fasterxml.woodstox:woodstox-core:jar:6.5.0:runtime
[INFO] |     \- org.codehaus.woodstox:stax2-api:jar:4.2.1:runtime
[INFO] +- org.eclipse.persistence:javax.persistence:jar:2.2.1:compile
[INFO] +- javax.transaction:jta:jar:1.1:compile
[INFO] +- org.primefaces:primefaces:jar:14.0.6:compile
[INFO] +- org.jasypt:jasypt:jar:1.9.3:compile
[INFO] +- net.sourceforge.jexcelapi:jxl:jar:2.6.12:compile
[INFO] +- com.mashape.unirest:unirest-java:jar:1.4.9:compile
[INFO] +- org.apache.httpcomponents:httpclient:jar:4.5.14:compile
[INFO] |  \- org.apache.httpcomponents:httpcore:jar:4.4.16:compile
[INFO] +- org.apache.httpcomponents:httpasyncclient:jar:4.1.5:compile
[INFO] |  \- org.apache.httpcomponents:httpcore-nio:jar:4.4.15:compile
[INFO] +- org.apache.httpcomponents:httpmime:jar:4.5.14:compile
[INFO] +- org.json:json:jar:20240303:compile
[INFO] +- net.glxn:qrgen:jar:1.4:compile
[INFO] |  \- com.google.zxing:javase:jar:3.0.0:compile
[INFO] +- net.sf.barcode4j:barcode4j-light:jar:2.0:compile
[INFO] |  \- xml-apis:xml-apis:jar:1.3.04:compile
[INFO] +- javax:javaee-endorsed-api:jar:7.0:compile
[INFO] +- org.keycloak:keycloak-common:jar:22.0.1:compile
[INFO] |  \- jakarta.activation:jakarta.activation-api:jar:2.1.2:compile
[INFO] +- javax:javaee-web-api:jar:8.0.1:compile
[INFO] +- org.apache.directory.studio:org.dom4j.dom4j:jar:1.6.1:compile
[INFO] |  \- dom4j:dom4j:jar:1.6.1:compile
[INFO] +- javax.mail:mail:jar:1.4.7:compile
[INFO] |  \- javax.activation:activation:jar:1.1:compile
[INFO] +- org.reflections:reflections:jar:0.10.2:compile
[INFO] |  \- org.javassist:javassist:jar:3.28.0-GA:compile
[INFO] +- org.apache.tomcat:tomcat-catalina:jar:9.0.83:compile
[INFO] |  +- org.apache.tomcat:tomcat-servlet-api:jar:9.0.83:compile
[INFO] |  +- org.apache.tomcat:tomcat-jsp-api:jar:9.0.83:compile
[INFO] |  |  \- org.apache.tomcat:tomcat-el-api:jar:9.0.83:compile
[INFO] |  +- org.apache.tomcat:tomcat-juli:jar:9.0.83:compile
[INFO] |  +- org.apache.tomcat:tomcat-annotations-api:jar:9.0.83:compile
[INFO] |  +- org.apache.tomcat:tomcat-api:jar:9.0.83:compile
[INFO] |  +- org.apache.tomcat:tomcat-jni:jar:9.0.83:compile
[INFO] |  +- org.apache.tomcat:tomcat-coyote:jar:9.0.83:compile
[INFO] |  +- org.apache.tomcat:tomcat-util:jar:9.0.83:compile
[INFO] |  +- org.apache.tomcat:tomcat-util-scan:jar:9.0.83:compile
[INFO] |  \- org.apache.tomcat:tomcat-jaspic-api:jar:9.0.83:compile
[INFO] +- com.github.hmislk:lims-middleware-libraries:jar:1.1.4:compile
[INFO] \- com.google.code.gson:gson:jar:2.8.9:compile

**Reply by melloware** [reply, marked as answer]
I see you have `com.github.librepdf:openpdf:jar:1.3.24:compile` ? Can you try the last 1.3.X which is `1.3.43`?

**Comment by adamgongca** [comment]
I have same issue with PrimeFaces 15.0.4

**Reply by melloware** [reply]
Please provide an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project.  It is the only way developers can debug your problem to help.
